src/coreclr/jit/objectalloc.cpp

```diff
@@ -1710,8 +1710,41 @@ unsigned int ObjectAllocator::MorphAllocObjNodeIntoStackAlloc(GenTreeAllocObj* a
     bool bbInALoop  = block->HasFlag(BBF_BACKWARD_JUMP);
     bool bbIsReturn = block->KindIs(BBJ_RETURN);
 
-    if (comp->fgVarNeedsExplicitZeroInit(lclNum, bbInALoop, bbIsReturn))
+    // We zero init locally if the allocation site is in a loop, a return, or
+    // (profitability heuristic) if the allocation site may execute after EH region.
+    // This latter prevents the promoted allocation fields from being EH live.
+    //
+    bool                    needsExplicitZeroInit = comp->fgVarNeedsExplicitZeroInit(lclNum, bbInALoop, bbIsReturn);
+    bool                    mayExecuteAfterEH     = false;
+    FlowGraphDfsTree* const dfsTree               = comp->m_dfsTree;
+
+    if (!needsExplicitZeroInit && (comp->compHndBBtabCount > 0) && (dfsTree->Contains(block)))
+    {
+        // Walk forward through RPO until we find the allocation block. This gives
+        // us a crude "may happen before" check. This does not need to be exact.
+        // Note we only do this if the allocation site is NOT in a loop.
+        //
+        for (unsigned i = dfsTree->GetPostOrderCount(); i != 0; i--)
+        {
+            BasicBlock* const rpoBlock = dfsTree->GetPostOrder(i - 1);
+
+            if (rpoBlock == block)
+            {
+                break;
+            }
+
+            if (comp->bbIsTryBeg(rpoBlock))
```

Comment:
```suggestion
            if (rpoBlock->hasTryIndex())
```

This check is slightly faster, and it holds that the first block we encounter in RPO with a try index has to be the try region's entry.

Reply:
Thanks -- let's do it that way.